Add FoldCase API

fbacher opened the pull request:
## Description
Adds FoldCase API, tests and several uses of the API.

Keeping commit small and focused on FoldCase itself. Followup commits will change numerous uses of ToLower (and similar code, all in StringUtils) to use FoldCase.

## Motivation and context
This change improves on patch for Turkish-I problem (19883). The Turkish-I problem caused Kodi to be non-functional due to ToLower not treating ASCII i/I and Turkish "I" characters in a "normal" manner. This caused all sorts of problems when the keys to lookup tables, etc. were normalized to lower case (or upper case).

Most of Kodi's use of ToLower is for 'normalizing' a key to a table. The vast majority of these cases use ASCII key values. But not universally. I have seen eight instances where something like "keyboard" was translated into the current language and then used as a key value. I have also seen where keywords are looked for in arbitrary text. These frequently work okay even with an inferior single-byte C++ tolower, but it is flirting with trouble.

This implementation uses Unicode case folding data from Unicode.org to perform "simple" casefolding. Simple case folding handles all cases where there is a 1-1 codepoint (Unicode "character") conversion between a codepoint and the folded version of the codepoint. It can't handle the few situations where a single codepoint (like German sharfes-S, which when folded becomes: 'ss' or where some context in the word or sentence makes a difference. For Kodi these situations should only exist if some quite poor programming practice is used for key values. Even if these do exist, implementing FoldCase would make the problem easier to fix since it separates general 'ToLower' (for the user to see) usage from 'FoldCase' (internal key) usage.
## How has this been tested?
For the first commit, the tests are only with the improved TestStringUtils.cpp. This is due to this commit being the API itself. The second and subsequent commits to use FoldCase will modify existing code but not additional tests. Additional tests will be added as additional APIs are added/modified (CompareNoCase, EqualsNoCase, etc.). I have worked on this problem for months, it has gone through multiple iterations and implementations. I have run all automated tests. I have run Kodi using English, German, Russian, Ukranian, Turkish and Japanese. I have downloaded several thousand movie titles from TMDb in a combination of the above languages and created dummy movies out of them so that I can search and play them. I have a Random Trailers plugin which runs in the background and plays the fake movies and displaying the Movie Info.

I have done the basic, easy things to do with the UI. I have navigated to numerous screens, etc. without failure. I have displayed our 25,000 (English) audio library.

That being said, it is very difficult to test more than a tiny amount of Kodi by hand. There are many features that I am not that familiar with. The problem is particularly difficult for me using non-English.

More testing that I have done:

    Sorting by title, date, etc.
    Search by actor, genre, date
    Drill down Movie Info -> Actor -> Titles (that the actor is in)

Testing on Ubuntu 24.04 with 6/12 CPU Intel with 32G ram. Usually running with debug builds, but also -O2 and with memory debugging from time to time.

## What is the effect on users?
Will prevent Turkish black-screen
As a side-effect the 'accuracy' of FoldCase, ToLower/ToUpper, and string comparisons should be improved for multi-byte languages since the current implementation mostly relies on C++ tolower/toupper, which ONLY works properly for single byte characters. This implementation of FoldCase is Unicode (4-bytes) ToLower/ToUpper will go at least to 2-bytes and probably 4-bytes (I'm looking at more character tables for ToLower/ToUpper from Unicode.org that have essentially the same format as for FoldCase).

Downside: Performance. Preliminary tests show that converting from utf-8 to utf32 and back can increase FoldCase string by 3x over current. However, this says nothing about real world. When I am finished converting all of the code over I will instrument the code to capture much more realistic data. There are also some things that I am doing which will help mitigate the problem, including the use of string_view instead of string.

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ x] I have added tests to cover my change
- [x ] All new and existing tests passed

fbacher committed:
Add FoldCase API

Adds FoldCase API, tests and several uses of the API.

Keeping commit small and focused on FoldCase itself.
Followup commits will change numerous uses of ToLower to use FoldCase

fbacher committed:
Use FoldCase for xbmc/utils

Switch all of the apprpriate calls to ToLower in xbmc/utils/... to use
    FoldCase instead

fbacher committed:
Use FoldCase for root, addons, application

Use FoldCase for: xbmc, xbmc/addons and xbmc/application; as appropriate
in place of ToLower

fbacher committed:
Use FoldCase in xbmc/cores

Use FoldCase instead of ToLower in the appropriate situations in the
xbmc/cores subdirectory

fbacher committed:
Use Foldcase for filesystem, games, gulib, input

Replace the use of ToLower with FoldCase in the appropriate places for
xbmc/filesystem, xbmc/games, xbmc/guilib, xbmc/input

fbacher committed:
Use FoldCase for settings, video, windowing, windows

Use FoldCase instead of ToLower for the appropriate instances in the
xbmc/settings, xbmc/video, xbmc/windowing and xbmc/windows directories.

fbacher committed:
Use FoldCase for interfaces, media, music, network, peripherals

Use FoldCase instead of ToLower in the appropriate situations for
xbmc/interfaces,
xbmc/media, xbmc/music, xbmc/network and xbmc/peripherals

garbear commented on xbmc/utils/test/TestStringUtils.h:
```diff
@@ -0,0 +1,595 @@
+/*
```
This is a good comment but could you add the standard copyright header before it?

fbacher replied:
Nice catch. Will do. Also, why is copyright to 2018 and not 2022?

garbear commented:
I gave the latest version another read-through and the code is looking good. I would only follow C++ traditions for headers and cpp files a little more closely.

fbacher commented on xbmc/utils/test/TestStringUtils.cpp:
```diff
@@ -6,11 +6,19 @@
  *  See LICENSES/README.md for more information.
  */
 
+#include "TestStringUtils.h"
```
I thought that putting these large, boring arrays in a .h to reduce clutter in the test itself.  But, I may as well try to be normal

fbacher committed:
Change *NoCase functions to use FoldCase

In addition, add some more functions where performance can be improved,
such as EqualsNoCase in addition to CompareNoCase.
Also, added Equals since a length-check can avoid the comparison.
Changed APIs to use string_view instead of the various permutations of
string and char*

fbacher committed:
Use Foldcase for pictures, platform, playlists, programs, pvr

Use Foldcase instead of ToLower in the appropriate cases for
xbmc/pictures, xbmc/platform, xbmc/playlists, xbmc/programs
xbmc/pvr

fbacher committed:
Private iconv for StringUtils

c++ wstring_convert does not convert to/from wstring & u32string, so
iconv needed.
CharSetConverter has signficant shortcomings:
- DOES NOT initialize prior to StringUtils needing it. Spent much time
trying to overcome. Seems that much surgery needed.
- Requires input to be strings rather than string_view
- Uses one converter of each time with locks preventing concurrent use.
Saw no reason to not use thread-local versions without locks
- mallocs scratch buffer, should use stack buffer

Implemented UnicodeConverter based on CharSetConverter to overcome these
issues.

fuzzard commented:
Are you needing this to be built by jenkins currently, or can I leave the no Jenkins label until you actually need full CI build coverage?

fbacher commented:
I'm trying to resolve Windows build issue. I can slow down how often I do these if it is becoming a resource problem. Perhaps I should upgrade my windows box and get a compiler so that I don't impact you so much.

garbear commented:
Hammer our CI. That's what it's there for. Jenkins can become a bit congested during European day hours, but as you can see [here](https://jenkins.kodi.tv) now it's totally unused.

fbacher committed:
Use new ToUpper/ToLower

ToUpper/ToLower need to be consistent with FoldCase, both with unicode
behavior and API.

Prior version had wstring characters treated differently than string
(used different character tables).

Now string & wstring versions perform ToLower using u32Char instead of a
byte or wchar_t at a time. Both now use latest Unicode.org simple case
UTF-32 tables. Results between ToLower, ToUpper & FoldCase should be
consistent in both string and wstring versions.

APIs now use string_view for input and return string/wstring results
instead of by reference. This was feasible since the number of uses of
these functions was greatly reduced after most changed to FoldCase.

fbacher committed:
Cleanup

Remove old APIs, dead code
General cleanup

fbacher commented:
superseded
